### PR TITLE
fix: address review comments from PRs #210/#211/#216/#217/#220

### PR DESCRIPTION
**Pull request by CybotTM**
## Summary
Consolidates the substantive fixes flagged by reviewers across this session's attachment-migration PRs. Non-substantive `except T, V:` style suggestions are explicitly not applied — see "Why not also..." section.

### #211 — `str(None) → "None"` validation gap
`str(None) == "None"` is truthy, silently bypassing the `if not filename or not url:` guard. Validate raw values **before** coercing in both `_map` and `_process_batch_end_to_end`.

### #210 / #217 — Ruby `File.open` FD leak
The Rails attach script's `file = File.open(path, 'rb')` left the descriptor open when `att.save!` raised, leaking FDs across hundreds of ops in a single `rails runner` invocation. Wrap in a block so it's released on every path.

### #217 — `_download_attachment` improvements
- Wrap `session.get(stream=True)` in a `with` block — return keep-alive socket on mid-stream errors.
- Stream-write to `<dest>.part` + atomic rename on success — never leave a partial file for the `local_path.exists()` check.
- Replace substring matching for HTTP 400 with `response.status_code` (fall through to substring only when no `response` on the exception, e.g. test doubles).

### #216 — Dotfile edge case + redundant import
- `[".bashrc", ".bashrc"] → [".bashrc", ".bashrc (2)"]` (was `" (2).bashrc"` with leading space + base lost).
- Drop the duplicate inner `from ...attachments_migration import AttachmentsMigration`.

### #220 — `cleanup_noname_op_attachments` improvements
- Pass project identifier via `input_data` instead of string-interpolating into Ruby.
- Single batched sibling-count query (was N+1).
- Dry-run summary reports `eligible_for_deletion` (always populated) — was `will_delete` (gated on apply), so the prior version always showed "0 eligible" in dry-run.
- CLI surfaces Ruby-side `error` payload to non-zero exit + stderr.
- `import json` at module top.

3 new unit tests pin the cleanup-tool fixes.

### `except T, V:` → `except (T, V):` (#214 review)
Python 3.14 [made comma-tuples valid in except clauses](https://docs.python.org/3.14/whatsnew/3.14.html); both forms produce the same `Tuple` AST. The project's `ruff format` (target-version py314) explicitly canonicalises the parenthesised form **back** to the bare form. Applying the suggested fix would revert on the next format run. The bare form is correct + idiomatic on 3.14 and the codebase uses it consistently.

### `inspect.getsource` test brittleness (#210/#211 review)
The alternative would require mocking the entire Rails round-trip just to pin a single string substring. Existing approach is intentional — it's a pin against accidental removal, not a behavioural assertion.

## Test plan
- [x] `pytest tests/unit -q -x` → 1505 passed.
- [x] `ruff check` + `ruff format --check` clean.
- [x] Manual: dry-run + `--apply` paths of cleanup tool work correctly.

### CODE REVIEW

File: tools/cleanup_noname_op_attachments.py [hunk 1]
```diff
@@ -121,27 +145,39 @@ def main(argv: list[str] | None = None) -> int:
         return 2
 
     op = OpenProjectClient()
-    script = _build_script(args.project_key, apply=args.apply)
-    envelope = op.execute_script_with_data(script, [])
+    script = _build_script(apply=args.apply)
+    envelope = op.execute_script_with_data(script, [{"identifier": args.project_key.lower()}])
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply tools/cleanup_noname_op_attachments.py:149]: Fixed in 44a6036 (https://github.com/netresearch/jira-to-openproject/pull/224). Real bug — caught it before any live re-run with the new data-passing path. The Ruby now reads `input_data.first['identifier']` when ``input_data`` is an Array, matching the runner's list-payload convention.

File: src/application/components/attachments_migration.py [hunk 2]
```diff
@@ -261,35 +261,68 @@ def _download_attachment(self, url: str, dest_path: Path) -> Path:
...
+                # sees ``not local_path.exists()`` and counts it under
+                # ``map_download_failed`` instead of misinterpreting
+                # an N-byte fragment as a successful download.
+                try:
+                    tmp_path.unlink(missing_ok=True)
+                except OSError:
+                    pass
+                raise
+
+        def _is_http_400(exc: BaseException) -> bool:
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply src/application/components/attachments_migration.py:303]: Fixed in 44a6036 (https://github.com/netresearch/jira-to-openproject/pull/224). `_is_http_400(exc: Exception)` — agreed, `BaseException` would also catch `KeyboardInterrupt` / `SystemExit` which we don't want to probe here.

File: tools/cleanup_noname_op_attachments.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (CybotTM) [reply tools/cleanup_noname_op_attachments.py:149]: Fixed in 44a6036 (https://github.com/netresearch/jira-to-openproject/pull/224). Same fix — read `input_data.first['identifier']` from the list payload.

File: src/application/components/attachments_migration.py [hunk 3]
```diff
@@ -335,11 +368,17 @@ def _map(self, extracted: ComponentResult) -> ComponentResult:
...
+                    # produces the truthy literal ``"None"``, which would
+                    # bypass the ``if not …`` guard and silently let an
+                    # invalid item through. Caught by PR #211 review.
+                    if not raw_filename or not raw_url:
                         self._loss_counters["map_missing_filename_or_url"] += 1
                         continue
+                    filename = str(raw_filename)
+                    url = str(raw_url)
                     safe_name = filename.replace("/", "_")
                     local_path = self.attachment_dir / safe_name
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply src/application/components/attachments_migration.py:383]: Fixed in 44a6036 (https://github.com/netresearch/jira-to-openproject/pull/224). Real bug. The local path now incorporates the Jira attachment id as a prefix (`f"{aid}_{safe_name}"`), so two attachments sharing a Jira filename can no longer overwrite each other on disk. The op's `filename` field still holds the original Jira filename so the Rails attach script writes the user-visible name correctly.

File: src/application/components/attachments_migration.py [hunk 4]
```diff
@@ -794,11 +844,14 @@ def _process_batch_end_to_end(
...
+                    raw_filename = item.get("filename")
+                    raw_url = item.get("url")
+                    # See ``_map`` — same ``str(None)`` gotcha.
+                    if not raw_filename or not raw_url:
                         self._loss_counters["map_missing_filename_or_url"] += 1
                         continue
+                    filename = str(raw_filename)
+                    url = str(raw_url)
                     safe_name = filename.replace("/", "_")
                     local_path = self.attachment_dir / safe_name
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply src/application/components/attachments_migration.py:856]: Fixed in 44a6036 (https://github.com/netresearch/jira-to-openproject/pull/224). Same per-attachment-id local-path scheme applied at the second site (`_process_batch_end_to_end`).
